<head>
<title>TRExplorer</title>
<link rel="icon" type="image/png" href="favicon.png">
<link rel="stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/semantic-ui/2.5.0/semantic.min.css">
<script src="https://code.jquery.com/jquery-3.7.1.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/semantic-ui/2.5.0/semantic.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/chart.js"></script>
<script src="https://cdn.jsdelivr.net/npm/chartjs-chart-matrix@2"></script>

<!-- script src="https://storage.googleapis.com/tgg-viewer/igv.js-bw2/dist/igv.min.js?version=2024-08-06"></script -->
<script src="https://cdn.jsdelivr.net/npm/igv@3.4.0/dist/igv.min.js"></script>

<!-- Google tag (gtag.js) -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-6R8FPBQRFH"></script>
<script>
window.dataLayer = window.dataLayer || [];
function gtag(){dataLayer.push(arguments);}
gtag('js', new Date());

gtag('config', 'G-6R8FPBQRFH');
</script>

{% include "styles_for_all_pages.html" %}

{% include "utils_for_dot_plot.html" %}
{% include "utils_for_igv.html" %}
{% include "utils_for_motif_frequency.html" %}
{% include "utils_for_periodicity_plot.html" %}
{% include "utils_for_polymorphism_charts.html" %}

<script>

    const TR_CATALOG_BASE_URL = "https://storage.googleapis.com/tandem-repeat-catalog"

    const SHARED_TR_CATALOG_SIZES = {
        known_disease_loci: 63,
        illumina_174k_catalog: 174300,
        tr_explorer_catalog: 4863041,
        tr_explorer_catalog_v2: 5543507,
    }

    const TR_EXPLORER_SOURCE_CATALOGS = [
        ["igv_kd_loci_july_2024", "Known disease-associated TR loci (July 2024)", `${TR_CATALOG_BASE_URL}/other_catalogs/KnownDiseaseAssociatedLoci_July2024.bed.gz`, SHARED_TR_CATALOG_SIZES['known_disease_loci']],
        ["igv_illumina_174k_source", "Illumina 174k", `${TR_CATALOG_BASE_URL}/other_catalogs/Illumina174kPolymorphicTRs.bed.gz`, SHARED_TR_CATALOG_SIZES['illumina_174k_catalog']],
        ["igv_perfect_repeats_hg38", "Perfect repeats in hg38", `${TR_CATALOG_BASE_URL}/other_catalogs/PerfectRepeatsInGRCh38.bed.gz`, 4558281],
        ["igv_polymorphic_trs_in_t2t", "Polymorphic TRs in T2T assemblies", `${TR_CATALOG_BASE_URL}/other_catalogs/PolymorphicTRsInT2TAssemblies.bed.gz`, 1937805],
    ]

    const TR_EXPLORER_CATALOG  = [
        ["igv_tr_explorer_catalog_v1", "TRExplorer catalog v1.0", `${TR_CATALOG_BASE_URL}/v1.0/TRExplorer.repeat_catalog_v1.hg38.1_to_1000bp_motifs.bed.gz`, SHARED_TR_CATALOG_SIZES['tr_explorer_catalog'], "TR catalog described in [<a href='https://www.biorxiv.org/content/10.1101/2024.10.04.615514v2' target=_blank>Weisburd, Dolzhenko et al. 2024</a>]"],
        ["igv_tr_explorer_catalog_v2", "TRExplorer catalog v2.0 draft", `${TR_CATALOG_BASE_URL}/v2.0/TRExplorer.repeat_catalog_v2.hg38.1_to_1000bp_motifs.bed.gz`, SHARED_TR_CATALOG_SIZES['tr_explorer_catalog_v2'], "Draft version 2.0 of the TR catalog described in [<a href='https://www.biorxiv.org/content/10.1101/2024.10.04.615514v2' target=_blank>Weisburd, Dolzhenko et al. 2024</a>]. This version introduces loci from the Vamos v2.1 catalog in order to improve coverage of VNTRs."],
    ]

    const TR_VARIATION_CLUSTERS = [
        ["igv_vc_v1", "Variation clusters v1.0", `${TR_CATALOG_BASE_URL}/v1.0/variation_clusters_v1.hg38.bed.gz`, 271362, "Variation Clusters (VCs) are genomic intervals that contain one or more TRs embedded within a wider region that harbors multiple common polymorphisms. For more details, see [<a href='https://www.biorxiv.org/content/10.1101/2024.10.04.615514v2' target=_blank>Weisburd, Dolzhenko et al. 2024</a>]."],
        ["igv_vc_and_isolated_v1", "Variation clusters and isolated repeats v1.0", `${TR_CATALOG_BASE_URL}/v1.0/TRExplorer.variation_clusters_and_isolated_TRs_v1.hg38.TRGT.bed.gz`, 4542828, "This catalog contains intervals representing variation clusters, as well as entries for all TRExplorer catalog loci that are not within a variation cluster. Variation Clusters (VCs) are genomic intervals that contain one or more TRs embedded within a wider region that harbors multiple common polymorphisms. For more details, see [<a href='https://www.biorxiv.org/content/10.1101/2024.10.04.615514v2' target=_blank>Weisburd, Dolzhenko et al. 2024</a>]."],
        ["igv_vc_and_isolated_v1_0_1", "Variation clusters and isolated repeats v1.0.1", `${TR_CATALOG_BASE_URL}/v1.0.1/TRExplorer.variation_clusters_and_isolated_TRs_v1.0.1.hg38.TRGT.bed.gz`, 4439813, "This catalog contains intervals representing variation clusters, as well as entries for all TRExplorer catalog loci that are not within a variation cluster. v1.0.1 removes 103,015 isolated repeat loci from v1.0 but that occur inside very large variation clusters which exceeded the max region size of the variation cluster algorithm. Variation Clusters (VCs) are genomic intervals that contain one or more TRs embedded within a wider region that harbors multiple common polymorphisms. For more details, see [<a href='https://www.biorxiv.org/content/10.1101/2024.10.04.615514v2' target=_blank>Weisburd, Dolzhenko et al. 2024</a>]."],
    ];

    const KNOWN_TR_LOCI_TRACK = [
        ["igv_kd_loci", "Disease-associated loci (March 2025)", `${TR_CATALOG_BASE_URL}/other_catalogs/KnownDiseaseAssociatedLoci.bed.gz`, 68],
        ["igv_trgt_kd_loci", "TRGT repo: disease-associated loci (March 2025)", `${TR_CATALOG_BASE_URL}/other_catalogs/TRGT.pathogenic_repeats.hg38.bed.gz`, 56],
        ["igv_strchive_kd_loci", "STRchive catalog for TRGT (March 2025)", `${TR_CATALOG_BASE_URL}/other_catalogs/STRchive-disease-loci-v2.4.0.hg38.TRGT.B2FLLAlV.bed.gz`, 73],
    ]

    const GENOME_WIDE_TR_CATALOGS = [
        ["igv_illumina_174k", "Illumina 174k", `${TR_CATALOG_BASE_URL}/other_catalogs/Illumina174kPolymorphicTRs.bed.gz`, SHARED_TR_CATALOG_SIZES['illumina_174k_catalog']],
        ["igv_trgt", "TRGT", `${TR_CATALOG_BASE_URL}/other_catalogs/TRGT.bed.gz`, 171146, "Genome-wide catalog provided in the <a href='https://github.com/PacificBiosciences/trgt/tree/3d10c75/repeats' target='_blank'>TRGT repo</a> (downloaded March, 2025)"],
        ["igv_ucsc_sr", "UCSC simple repeat track", `${TR_CATALOG_BASE_URL}/other_catalogs/UCSC_SimpleRepeatTrack.bed.gz`, 965511],
        ["igv_vamos", "VAMOS v2.1", `${TR_CATALOG_BASE_URL}/other_catalogs/VamosCatalog_v2.1.bed.gz`, 1243954],
        ["igv_gangstr", "GangSTR v17", `${TR_CATALOG_BASE_URL}/other_catalogs/GangSTR_v17.bed.gz`, 1340266],
        ["igv_hipstr", "HipSTR", `${TR_CATALOG_BASE_URL}/other_catalogs/HipSTR.bed.gz`, 1638945],
        ["igv_adotto", "Adotto v1.2", `${TR_CATALOG_BASE_URL}/other_catalogs/Adotto_v1.2.bed.gz`, 2934805],
        ["igv_popstr", "PopSTR v2", `${TR_CATALOG_BASE_URL}/other_catalogs/PopSTR_v2.bed.gz`, 5401401],
        ["igv_platinum_trs", "Platinum TRs", `${TR_CATALOG_BASE_URL}/other_catalogs/PlatinumTRs.bed.gz`, 7524815, "The TR catalog described in [<a href='https://pubmed.ncbi.nlm.nih.gov/39149261/' target=_blank>Porubsky et al. 2024</a>]"],
        ["igv_chiu", "Chiu et al.", `${TR_CATALOG_BASE_URL}/other_catalogs/Chiu_et_al.bed.gz`, 18759399, "The TR catalog described in [<a href='https://pubmed.ncbi.nlm.nih.gov/38947075/' target=_blank>Chiu et al. 2024</a>]"],
    ]

    REFERENCE_BASE_URL = "https://storage.googleapis.com/tgg-viewer/ref/GRCh38"

    const REFERENCE_TRACKS = [
        ["igv_gencode", "Gencode v49", `${REFERENCE_BASE_URL}/gencode_v49/gencode.v49.GRCh38.sorted.txt.gz`],
        ["igv_segdups", "SegDups > 1000bp", `${REFERENCE_BASE_URL}/segdups/segdups.gtf.gz`],
        ["igv_36k_map", "36-mer mappability", `${REFERENCE_BASE_URL}/mappability/GRCh38_no_alt_analysis_set_GCA_000001405.15-k36_m2.bw`],
        ["igv_100k_map", "100-mer mappability", `${REFERENCE_BASE_URL}/mappability/GRCh38_no_alt_analysis_set_GCA_000001405.15-k100_m2.bw`],
        ["igv_haploinsuf", "ClinGen haploinsufficiency track", `${REFERENCE_BASE_URL}/clingen/ClinGen_haploinsufficiency_gene_GRCh38.sorted.bed.gz`],
        ["igv_triplosens", "ClinGen triplosensitivity track", `${REFERENCE_BASE_URL}/clingen/ClinGen_triplosensitivity_gene_GRCh38.sorted.bed.gz`],
        ["igv_repeatmask", "UCSC RepeatMasker", `${REFERENCE_BASE_URL}/hg38.RepeatMasker.bed.gz`],
    ]

    const ALL_AVAILABLE_TRACKS = [
        ...TR_EXPLORER_SOURCE_CATALOGS,
        ...TR_EXPLORER_CATALOG,
        ...TR_VARIATION_CLUSTERS,
        ...GENOME_WIDE_TR_CATALOGS,
        ...KNOWN_TR_LOCI_TRACK,
        ...REFERENCE_TRACKS,
    ]

    // initialize BigQuery client
    const PROJECT_ID = 'cmg-analysis'
    const DATASET_ID = 'tandem_repeat_explorer'
    const TABLE_ID = 'catalog_20251117_063426'
    const COMPLETE_TABLE_ID = `\`${PROJECT_ID}.${DATASET_ID}.${TABLE_ID}\``

    const SHOW_COLUMNS_VALUE_DELIMITER = 'i'

    const IGV_LOCUS_PADDING = 150
    const REFERENCE_REGION_REGEXP = /^(?:chr)?([XYMTt0-9]+):([\d,]+)-([\d,]+)$/i

    const ALL_STDEV_COLUMNS = [
        'HPRC100_Stdev',
        'AoU1027_Stdev',
        'TenK10K_Stdev',
        'StdevFromIllumina174k',
        'StdevFromT2TAssemblies',
    ]

    const STDEV_SIGMA_SUBSCRIPTS = {
        'HPRC100_Stdev': 'Hprc100',
        'AoU1027_Stdev': 'AoU1027',
        'TenK10K_Stdev': '10k10k',
        'StdevFromIllumina174k': '1kGP',
        'StdevFromT2TAssemblies': 'T2T',
    }

    const AXIS_UNITS = {
        'HPRC100_AlleleHistogram': 'Repeat Count: LPS',
        'TenK10K_AlleleHistogram': 'Repeat Count',
        'AlleleFrequenciesFromIllumina174k': 'Repeat Count',
        'AlleleFrequenciesFromT2TAssemblies': 'Repeat Count',
        'HPRC100_BiallelicHistogram': 'Repeat Count: LPS',
        'TenK10K_BiallelicHistogram': 'Repeat Count',
    }

    const ALLELE_HISTOGRAM_SHORT_NAMES = {
        'HPRC100_AlleleHistogram': 'HPRC100',
        'TenK10K_AlleleHistogram': 'TenK10K Phase 1',
        'AlleleFrequenciesFromIllumina174k': '1kGP',
        'AlleleFrequenciesFromT2TAssemblies': 'T2T',
        'HPRC100_BiallelicHistogram': 'HPRC100',
        'TenK10K_BiallelicHistogram': 'TenK10K Phase 1',
    }
    const ALLELE_FREQUENCY_CHART_TITLES = {
        'HPRC100_AlleleHistogram': 'Allele Frequencies from HPRC 100 PacBio samples',
        'TenK10K_AlleleHistogram': 'Allele Frequencies from TenK10K Phase 1 short-read genomes',
        'AlleleFrequenciesFromIllumina174k': 'Allele Frequencies from 1kGP short-read genomes',
        'AlleleFrequenciesFromT2TAssemblies': 'Allele Frequencies from 78 diploid T2T assemblies',
        'HPRC100_BiallelicHistogram': 'Genotype Heatmap from HPRC 100 PacBio samples',
        'TenK10K_BiallelicHistogram': 'Genotype Heatmap from TenK10K Phase 1 short-read genomes',
    }

    const ALLELE_FREQUENCY_CHART_SUBTITLES = {
        'HPRC100_AlleleHistogram': 'based on TRGT calls in 100 PacBio HiFi samples from the Human Pan-genome Reference Consortium (HPRC)',
        'TenK10K_AlleleHistogram': 'based on ExpansionHunter calls in 1,925 short-read genomes from the <a href=https://www.biorxiv.org/content/10.1101/2024.11.02.621562v2 target=_blank>TenK10K Phase 1 dataset</a>',
        'AlleleFrequenciesFromIllumina174k': 'based on ExpansionHunter calls in <a href=https://github.com/Illumina/RepeatCatalogs/blob/master/docs/str_diversity_1kg.md target=_blank>2,504 short-read genomes from 1kGP</a>',
        'AlleleFrequenciesFromT2TAssemblies': 'based on assembly-to-hg38 alignments of 78 diploid T2T assemblies as described in <a href=https://pubmed.ncbi.nlm.nih.gov/37214979 target=_blank>Weisburd et al. 2023</a>',
        'HPRC100_BiallelicHistogram': 'based on TRGT calls in 100 PacBio HiFi samples from the Human Pan-genome Reference Consortium (HPRC)',
        'TenK10K_BiallelicHistogram': 'based on ExpansionHunter calls in 1,925 short-read genomes from the <a href=https://www.biorxiv.org/content/10.1101/2024.11.02.621562v2 target=_blank>TenK10K Phase 1 dataset</a>',
    }

    const STDEV_HELP_TEXTS = {
        'HPRC100_Stdev': `This is the standard deviation of the allele frequency distribution ${ALLELE_FREQUENCY_CHART_SUBTITLES['HPRC100_AlleleHistogram']}. Allele sizes were computed using the longest pure segment (LPS) as described in [<a href=&quot;https://www.biorxiv.org/content/10.1101/2025.01.06.631535v2&quot; target=_blank>Danzi et al. 2025</a>]. <br /><br /><b>Range:</b> 0 to 330.7`,
        'AoU1027_Stdev': `This is the standard deviation of the allele frequency distribution based on TRGT calls in PacBio HiFi samples from 1,027 African American participants in the All of Us (AoU) project. Allele sizes were computed using the longest pure segment (LPS) as described in [<a href=&quot;https://www.biorxiv.org/content/10.1101/2025.01.06.631535v2&quot; target=_blank>Danzi et al. 2025</a>]. <br /><br /><b>Range:</b> 0 to 932.1`,
        'TenK10K_Stdev': `This is the standard deviation of the allele frequency distribution ${ALLELE_FREQUENCY_CHART_SUBTITLES['TenK10K_AlleleHistogram']}. <br /><br />TenK10K Phase 1 primarily contains individuals of European ancestry.`,
        'StdevFromIllumina174k': `This is the standard deviation of the allele frequency distribution ${ALLELE_FREQUENCY_CHART_SUBTITLES['AlleleFrequenciesFromIllumina174k']}. <br /><br /><b>Range:</b> 0 to 31.5`,
        'StdevFromT2TAssemblies': `This is the standard deviation of the allele frequency distribution ${ALLELE_FREQUENCY_CHART_SUBTITLES['AlleleFrequenciesFromT2TAssemblies']}. <br /><br /><b>Range:</b> 0 to 1267.3`
    }

    const ALLELE_HISTOGRAM_HELP_TEXTS = {
        'HPRC100_AlleleHistogram': `Allele frequencies ${ALLELE_FREQUENCY_CHART_SUBTITLES['HPRC100_AlleleHistogram']}.`,
        'TenK10K_AlleleHistogram': `Allele frequencies ${ALLELE_FREQUENCY_CHART_SUBTITLES['TenK10K_AlleleHistogram']}.`,
        'AlleleFrequenciesFromIllumina174k': `Allele frequencies ${ALLELE_FREQUENCY_CHART_SUBTITLES['AlleleFrequenciesFromIllumina174k']}.`,
        'AlleleFrequenciesFromT2TAssemblies': `Allele frequencies ${ALLELE_FREQUENCY_CHART_SUBTITLES['AlleleFrequenciesFromT2TAssemblies']}.`,
        'HPRC100_BiallelicHistogram': `Allele frequencies ${ALLELE_FREQUENCY_CHART_SUBTITLES['HPRC100_BiallelicHistogram']}.`,
        'TenK10K_BiallelicHistogram': `Allele frequencies ${ALLELE_FREQUENCY_CHART_SUBTITLES['TenK10K_BiallelicHistogram']}.`,
    }

    const gene_region_priority = "with significance defined as coding > 5' UTR > 3' UTR > exon in non-coding gene > intron > promoter region spanning 1,000bp upstream of the 5' UTR"
    const RESULTS_TABLE_COLUMN_HELP_TEXTS = {
        LocusId: 'A unique id of this TR locus in hg38',
        ReferenceRegion: 'Genomic coordinates of the TR locus in the hg38 reference genome. <b>Example:</b> 1:94418421-94418442',
        ReferenceRepeatSequence: 'The repeat sequence in hg38',
        ReferenceLeftFlank: 'The sequence immediately to the left of the TR locus in hg38',
        ReferenceRightFlank: 'The sequence immediately to the right of the TR locus in hg38',
        ReferenceMotif: 'The repeat motif in hg38 and its size in base pairs',
        CanonicalMotif: 'The repeat motif in hg38, normalized by computing all cyclic rotations of the motif and its reverse complement, and then selecting the one that is alphabetically first. <b>For example:</b> the canonical version of <i>CTG</i> is <i>AGC</i>',
        polymorphism: 'Summary of available polymorphism data for the TR locus',
        Source: 'TRExplorer catalog version, as well as the source catalog that contributed this TR locus to the TRExplorer catalog',
        TRsInRegion: 'Number of other TRs in the vicinity of this locus that are separated from each other by no more than 6bp of spacer sequence',
        NumRepeatsInReference: 'Number of repeats calculated by taking the width of the TR locus reference interval and dividing it by the motif size, then rounding down to the nearest integer',
        ReferenceRepeatPurity: 'The fraction of bases within the TR locus reference sequence that match perfect repeats of the specified motif. For example, CAG.CAG.CAG has purity = 1.0, while CAG.CAA.CAG has purity = 0.89<br /><br /><b>Range:</b> 0 to 1',
        variationCluster: 'Summarizes whether the TR locus resides within a variation cluster. Variation Clusters (VCs) are genomic intervals that contain one or more TRs embedded within a wider region that harbors multiple common polymorphisms. For more details, see [<a href=\'https://www.biorxiv.org/content/10.1101/2024.10.04.615514v2\' target=_blank>Weisburd, Dolzhenko et al. 2024</a>]',
        isPathogenic: 'Whether this is a known disease-associated locus (provided as a separate column to enable sorting)',
        AoU1027_OE_LengthPercentile: 'Observed / Expected TR length metric from the TR constraint model described in [<a href=&quot;https://www.biorxiv.org/content/10.1101/2025.01.06.631535v2&quot; target=&quot;_blank&quot;>Danzi et al. 2025</a>], displayed as a percentile rank relative to all other TR loci in the catalog',
        RepeatMaskerIntervals: 'UCSC Repeat masker intervals that overlap the TR locus',
        VamosMotifFrequencies: 'Different motifs detected at this locus by the Vamos v2.1 pipeline when applied to 94 HPRC assemblies',
        FlanksAndLocusMappability: 'UCSC 36-mer mappability scores averaged across the TR locus and +/- 150bp of its flanking sequences<br /><br /><b>Range:</b> 0 to 1',
        GencodeGeneRegion: `The most significant gene region that overlaps the TR locus in Gencode v48 ${gene_region_priority}`,
        GencodeGeneName: 'Gencode v48 gene name',
        GencodeGeneId: 'Gencode v48 gene ID (ENSG)',
        GencodeTranscriptId: 'Gencode v48 transcript ID (ENST)',
        RefSeqGeneRegion: `The most significant RefSeq gene region that overlaps the TR locus ${gene_region_priority}`,
        RefSeqGeneId: 'RefSeq gene ID',
        RefSeqTranscriptId: 'RefSeq transcript ID',
        RefSeqGeneName: 'RefSeq gene name',
        ManeGeneRegion: `The most significant gene region that overlaps the TR locus in MANE v1.4 ${gene_region_priority}`,
        ManeGeneId: 'MANE v1.4 gene ID',
        ManeGeneName: 'MANE v1.4 gene name',
        ManeTranscriptId: 'MANE v1.4 transcript ID',
        HPRC100_BiallelicHistogram: 'Biallelic histogram of HPRC100',
        TenK10K_BiallelicHistogram: 'Biallelic histogram of TenK10K Phase 1',
        dotPlot: 'Dot plot visualization of the reference repeat sequence and flanking sequences. Vertical blue lines represent locus boundaries',
    }

    /*
    const GENE_REGION_COLORS = {
        'CDS': 'rgb(208,5,5)',
        'exon': 'rgb(255,168,0)',
        'intron': 'rgb(40,126,40)',
        'intergenic': 'rgb(0,178,255)',
        '5\' UTR': 'rgb(0,72,255)',
        '3\' UTR': 'rgb(0,72,255)',
        'promoter': 'rgb(93,1,196)',
    }
    */

    const DEFAULT_SORT_COLUMN = 'chrom_index, start_0based'
    const DEFAULT_SORT_DIRECTION = 'ASC'


    // persistent page state (added to the URL)
    let DEFAULT_GLOBAL_PAGE_STATE = {
        currentPage: 1,
        pageSize: 100,

        sc: DEFAULT_SORT_COLUMN,
        sd: DEFAULT_SORT_DIRECTION,

        showIGV: 0,
        igvLoc: 'chr4:3074821-3074990',

        showRs: 0,  //show results table
        searchQuery: '',
        showAdvFilters: 0,

        sourceCatalog: 'tr_explorer_catalog_v2', // Source catalog filter
        keepMotifSize: '',
        keepMotif: '',

        keepCodingLoci: 0,
        keepFiveUTRLoci: 0,
        keepThreeUTRLoci: 0,
        keepPromoterLoci: 0,
        keepIntronLoci: 0,
        keepIntergenicLoci: 0,
        keepNonCodingExonLoci: 0,

        polymorphismFilter: 0,
        polymorphismFilterType: '',
        polymorphismThreshold: '',
        vcFilter: 0,
        vcFilterType: '',

        // Custom track URLs
        igv_custom_track_url_1: '',
        igv_custom_track_url_2: '',
        igv_custom_track_url_3: '',
    }

    // set default IGV.js tracks
    ALL_AVAILABLE_TRACKS.forEach((rt) => {
        DEFAULT_GLOBAL_PAGE_STATE[rt[0]] = 0
    })
    DEFAULT_GLOBAL_PAGE_STATE['igv_kd_loci'] = 1
    DEFAULT_GLOBAL_PAGE_STATE['igv_tr_explorer_catalog_v2'] = 1


    // based on https://raw.githubusercontent.com/broadinstitute/str-analysis/refs/heads/main/str_analysis/variant_catalogs/variant_catalog_without_offtargets.GRCh38.json
    const MOTIFS_AT_KNOWN_DISEASE_ASSOCIATED_LOCI =  [
        'CAG',
        'GCC',
        'GAA',
        'GTC',
        'CAGG',
        'TTTG',
        'GGGCC',
        'ATTCT',
        'GGCCCC',
        'GGCCTG',
        'GGCGCGGAGC',
        'CGCGGGGCGGGG',
        'CCTCGCTGTGCCGCTGCCGA',
        'CCTCATGGTGGTGGCTGGGGGCAG',

        /*
        //BEAN1
        "TAGAA",
        "TGGAA",

        //DAB1
        "GAAAT",

        //MARCHF6
        "ATTTC",

        //RAPGEF2,
        "TTTCA",

        //RFC1
        "AAGGC",
        "AAGGG",
        "ACAGG",
        "AGAGG",
        "AGGGC",
        */
    ].join(', ')


    const writePersistentPageStateToUrl = () => {
        // Create object with only non-default values
        const nonDefaultParams = {}
        for (const [key, value] of Object.entries(GLOBAL_PAGE_STATE)) {
            if (value !== DEFAULT_GLOBAL_PAGE_STATE[key]) {
                nonDefaultParams[key] = value
            }
        }

        // Convert to URL hash using jQuery's param method
        const hash = '#' + $.param(nonDefaultParams)

        // Get current hash to check if state has changed
        const currentHash = window.location.hash
        if (currentHash !== hash) {
            // Update URL and add to browser history
            window.history.pushState(null, '', hash)
        }
    }

    const computeIntervalSize = (interval) => {
        const [chrom, start_end] = interval.split(':')
        const [start, end] = start_end.split('-')
        return parseInt(end) - parseInt(start)
    }


    const queryBigQuery = async (query, startIndex=null, pageSize=null, exportToFileFormat=null, toolName=null) => {
        try {
            const url = `https://us-central1-cmg-analysis.cloudfunctions.net/query_db`
            console.log("Page state:", GLOBAL_PAGE_STATE)
            console.log(exportToFileFormat ? `Exporting to ${toolName || ''} ${exportToFileFormat}:`: `Sending query:`, query.replace(/[\n\r\s]+/g, ' '))
            const requestBody = {
                sql: query,
                export_to_file_format: exportToFileFormat,
                tool_name: toolName,
                start_index: startIndex,
                page_size: pageSize,
            }

            const response = await fetch(url, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify(requestBody),
            })

            if (!response.ok) {
                throw new Error(`BigQuery API error: ${response.status} ${response.statusText}`)
            }

            return await response.json()
        } catch (error) {
            console.error('Error querying BigQuery:', error)
            throw new Error(`BigQuery query error: ${error.message || 'Unknown error'}`)
        }
    }

    const parseAlleleFrequenciesString = (frequenciesString) => {
        try {
            const data = []
            const items = frequenciesString.split(',')
            for (let i = 0; i < items.length; i++) {
                const [size, count] = items[i].split(':')
                data.push({
                    size: parseInt(size.replace('x', '')),
                    count: parseInt(count)
                })
            }
            return data
        } catch (e) {
            console.error('Error parsing allele frequencies string:', e)
            return []
        }
    }

    const parseRepeatMaskerIntervals = (repeatMaskerIntervals) => {
        if (!repeatMaskerIntervals) {
            return ''
        }

        let values = new Set()
        let count = 0
        let details = []
        for (let token of repeatMaskerIntervals.split(', ')) {
            const tokenEntries = token.split(':')
            if (!'Simple_repeat|Low_complexity'.includes(tokenEntries[0])) {
                values.add(tokenEntries[0])
                count += 1
                details.push(`<span style='font-family: monospace'>${tokenEntries[0]}: ${tokenEntries[1]}</span>`)
            }
        }

        if (count == 0) {
            return ''
        }

        return `<span class="show-popup action-link" data-html="${details.join('<br />')}">${Array.from(values).join(', ')}</span>`
    }

    const showTruncatedSequence = (sequence, truncateAt=12, showSize=true, leftTruncate=false, addBpSuffix=false) => {
        let sequenceSizeString = ''
        if (showSize) {
            sequenceSizeString += `(${sequence.length}${addBpSuffix? 'bp' : ''})`
        }
        if (sequence.length > truncateAt) {
            let result = `<span class="show-popup action-link" style="white-space: nowrap" data-html="<span style='font-family: monospace'>${sequence.replace(/.{1,100}/g, chunk => chunk + ' ')}</span>">`
            if (leftTruncate) {
                const truncatedSequence = sequence.substring(sequence.length - truncateAt, sequence.length)
                result += `<span style="font-family: monospace">${sequenceSizeString}</span> &nbsp; ...${truncatedSequence}`
            } else {
                const truncatedSequence = sequence.substring(0, truncateAt)
                result += `<span style="font-family: monospace">${truncatedSequence}</span>... &nbsp; ${sequenceSizeString}`
            }
            result += `</span>`
            return result
        } else {
            let result = `<span style="white-space: nowrap">`
            if (leftTruncate) {
                result += `${sequenceSizeString} &nbsp; <span style="font-family: monospace">${sequence}</span>`
            } else {
                result += `<span style="font-family: monospace">${sequence}</span> &nbsp; ${sequenceSizeString}`
            }
            result += `</span>`
            return result
        }
    }

    const computeMotifFrequenciesTableRows = (frequenciesString) => {
        if (!frequenciesString) {
            return []
        }
        const items = frequenciesString.split(',')
        let total = 0
        let motifCounts = {}
        let efficientMotifs = {}
        for (let i = 0; i < items.length; i++) {
            let motif = '', efficientMotif = '', count = '0'
            const numTokens = items[i].split(':').length
            if (numTokens == 3) {
                [motif, efficientMotif, count] = items[i].split(':')
            } else if(numTokens == 2) {
                [motif, count] = items[i].split(':')
            }

            count = parseInt(count)
            if (Number.isNaN(count)) {
                console.warn(`Unable to parse the count for motif ${motif} in ${frequenciesString}. Skipping...`)
                continue
            }
            motif = motif.trim()
            motifCounts[motif] = count
            efficientMotifs[motif] = efficientMotif.trim()
            total += count
        }

        const minimumFrequencyThreshold = 0  // 0.01

        // convert counts to fractions
        let motifCount = 0
        let homopolymerMotifCount = 0
        for (let motif in motifCounts) {
            motifCounts[motif] /= total
            if (motifCounts[motif] >= minimumFrequencyThreshold) {
                motifCount += 1
                if (motif.length == 1) {
                    homopolymerMotifCount += 1
                }
            }
        }

        if (motifCount <= 1 || homopolymerMotifCount == motifCount) {
            return []
        }

        const truncateMotifAt = 35
        let resultMotifs = []
        for (let motif in motifCounts) {
            let motifRow
            let motifPercentString
            if (motifCounts[motif] >= 0.01) {
                motifRow = '<tr>'
                motifPercentString = (motifCounts[motif] * 100.0).toFixed(0)
            } else {
                motifRow = `<tr style='color:#AAAAAA'>`
                motifPercentString = (motifCounts[motif] * 100.0).toFixed(1)
            }
            motifRow += `<td style='text-align:right'><i>${motifPercentString}%</i></td>`
            motifRow += `<td><span style='white-space: nowrap; font-family: monospace'>${motif.length > truncateMotifAt ? (motif.substring(0, truncateMotifAt) + '... (' + motif.length + 'bp)') : motif}</span></td>`
            if (efficientMotifs[motif]) {
                motifRow += `<td><span style='white-space: nowrap; font-family: monospace'>`
                if (efficientMotifs[motif] != motif) {
                    motifRow += `${efficientMotifs[motif].length <= truncateMotifAt ? efficientMotifs[motif] : (efficientMotifs[motif].substring(0, truncateMotifAt) + '... (' + efficientMotifs[motif].length + 'bp)')}`
                } else {
                    motifRow += '<i>same as motif</i>'
                }
                motifRow += `</span></td>`
            } else {
                motifRow += `<td><i>unavailable</i></td>`
            }
            motifRow += '</tr>'

            resultMotifs.push(motifRow)
        }

        return resultMotifs
    }

    const getGeneRegionColumnValue = (geneRegion) => {
        if (!geneRegion) return ''
        //const color = GENE_REGION_COLORS[geneRegion]
        const regionRename = {
            'CDS': 'coding',
            'exon': 'exon (non-coding gene)'
        }
        return `${regionRename[geneRegion] || geneRegion}`
    }

    const getPolymorphismSummaryColumnValue = (row) => {
        return `<span class="cell-in-polymorphism-column" ${getHtmlAttributesForPolymorphism(row)}>
            <span class="action-link">
                <b>σ<sub>${STDEV_SIGMA_SUBSCRIPTS['HPRC100_Stdev']}</sub></b>
            </span>
            <span style="color: #333;">
                = ${row.HPRC100_Stdev != null ? row.HPRC100_Stdev.toFixed(1) : '<span style="color: #888888;">not available</span>'} &nbsp; &nbsp;
            </span>
            ${row.AoU1027_Stdev != null ? `<span class="action-link">σ<sub>${STDEV_SIGMA_SUBSCRIPTS["AoU1027_Stdev"]}</sub></span>` : ''}
            ${row.AoU1027_Stdev != null ? `<span style="color: #333;">
                = ${row.AoU1027_Stdev != null ? row.AoU1027_Stdev.toFixed(1) : '<span style="color: #888888;">not available</span>'} &nbsp; &nbsp;
            </span>` : ''}
            <span class="action-link">
                ${row['TenK10K_Stdev'] != null ? `σ<sub>${STDEV_SIGMA_SUBSCRIPTS["TenK10K_Stdev"]}</sub>` : ''}  ${row.StdevFromIllumina174k != null ? `σ<sub>${STDEV_SIGMA_SUBSCRIPTS["StdevFromIllumina174k"]}</sub>` : ''}  ${row.StdevFromT2TAssemblies != null ? `σ<sub>${STDEV_SIGMA_SUBSCRIPTS["StdevFromT2TAssemblies"]}</sub>` : ''}
            </span>
        </span>
        `
    }

    const generatePolymorphismDialogContent = (row) => {
        const stdevColumns = ['HPRC100_Stdev', 'AoU1027_Stdev', 'TenK10K_Stdev', 'StdevFromIllumina174k', 'StdevFromT2TAssemblies']
        const alleleFrequencyColumns = ['HPRC100_AlleleHistogram', 'TenK10K_AlleleHistogram', 'AlleleFrequenciesFromIllumina174k', 'AlleleFrequenciesFromT2TAssemblies']

        const nStdevsNotAvailable = [row.HPRC100_Stdev == null, row.AoU1027_Stdev == null, row.TenK10K_Stdev == null, row.StdevFromIllumina174k == null, row.StdevFromT2TAssemblies == null].filter(Boolean).length

        let polymorphismDialogContent1 = `<div class="content info-font-size" style="text-align: center">`

        if (nStdevsNotAvailable == stdevColumns.length) {
            polymorphismDialogContent1 += `<span style="color: #686868;">No polymorphism data available</span><br /><br />`
        } else {
            polymorphismDialogContent1 += `
                <div><span style="font-weight: bold">Allele Frequency Distribution Summary Statistics</span></div>
                    <div style="margin-top: 5px">
                        <table style="margin: 0 auto">
            `
        }

        for (const key of stdevColumns) {
            if (row[key] != null) {
                polymorphismDialogContent1 += `
                    <tr>
                        <td style="text-align: left;">
                            <span style="font-size: 1.2em;">σ<sub style="font-size: 0.7em;">${STDEV_SIGMA_SUBSCRIPTS[key]}</sub></span>
                        </td>
                        <td style="text-align: right;">
                            = ${row[key].toFixed(3)}
                        </td>
                        <td style="text-align: right;">
                            <i class="question circle icon link" style="margin-left: 15px;" data-html="${STDEV_HELP_TEXTS[key]}"></i>
                        </td>
                    </tr>
                `
            }
        }
        polymorphismDialogContent1 += "</table>"

        if (nStdevsNotAvailable > 0) {
            polymorphismDialogContent1 += "<br />"
            let namesOfMissingSigmas = []
            for (const key of stdevColumns) {
                if (row[key] == null) {
                    namesOfMissingSigmas.push(`<span style="font-size: 1.2em;">σ<sub style="font-size: 0.7em;">${STDEV_SIGMA_SUBSCRIPTS[key]}</sub></span>`)
                }
            }
            polymorphismDialogContent1 += ` ${namesOfMissingSigmas.join(' &amp; ')} ${namesOfMissingSigmas.length > 1 ? 'are' : 'is'} not available for this locus`
        }

        if (row.AoU1027_OE_LengthPercentile != null) {
            polymorphismDialogContent1 += `
                <div style="margin-top: 25px">
                    Constraint (O/E Length) percentile = ${((100*row.AoU1027_OE_LengthPercentile).toFixed(1))}% <i class="question circle icon link" style="margin-left: 15px;" data-html="${RESULTS_TABLE_COLUMN_HELP_TEXTS['AoU1027_OE_LengthPercentile']}"></i>
                </div>
            `
        }

        polymorphismDialogContent1 += "</div>"

        // check if any of the allele frequency columns are available
        const alleleFrequencyColumnsAvailable = alleleFrequencyColumns.some(column => row[column] != null)
        if (alleleFrequencyColumnsAvailable) {
            // calculate the height of the allele frequency charts
            polymorphismDialogContent1 += `
                <div style="margin-top: 15px;">
                    <div style="display: flex; flex-direction: column; gap: 10px; margin-top: 10px">
            `

            for (const columnName of alleleFrequencyColumns) {
                if (row[columnName] != null) {
                    polymorphismDialogContent1 += `
                        <div style="border-top: 1px dashed #ccc; padding: 23px 20px 15px 15px;">
                            <div style="margin-bottom: 5px;"><b>${ALLELE_FREQUENCY_CHART_TITLES[columnName]}</b><br /><span style="color: #686868;">${ALLELE_FREQUENCY_CHART_SUBTITLES[columnName]}</span></div>
                            <canvas class="allele-frequency-chart" data-label="${ALLELE_HISTOGRAM_SHORT_NAMES[columnName]}" data-frequencies-column="${columnName}" data-ref-allele-size="${row.NumRepeatsInReference}" data-axis-units="${AXIS_UNITS[columnName]}" data-chart-width="500px" data-chart-height="200px" style="width: 100%; height: 100px;"></canvas>
                        </div>
                    `
                }
            }
            polymorphismDialogContent1 += '</div></div>'
        }

        const biallelicHistogramColumns = ['HPRC100_BiallelicHistogram', 'TenK10K_BiallelicHistogram']
        const biallelicHistogramColumnsAvailable = biallelicHistogramColumns.some(column => row[column] != null)
        if (biallelicHistogramColumnsAvailable) {
            polymorphismDialogContent1 += `
                <div style="margin-top: 15px;">
                    <div style="display: flex; flex-direction: column; gap: 10px; margin-top: 10px">
            `
            for (const columnName of biallelicHistogramColumns) {
                if (row[columnName] != null) {
                    polymorphismDialogContent1 += `
                        <div style="border-top: 1px dashed #ccc; padding: 23px 20px 15px 15px;">
                            <div style="margin-bottom: 5px;"><b>${ALLELE_FREQUENCY_CHART_TITLES[columnName]}</b><br /><span style="color: #686868;">${ALLELE_FREQUENCY_CHART_SUBTITLES[columnName]}</span></div>
                            <canvas class="biallelic-histogram-chart" data-label="${ALLELE_HISTOGRAM_SHORT_NAMES[columnName]}" data-frequencies-column="${columnName}" data-ref-allele-size="${row.NumRepeatsInReference}" data-axis-units="${AXIS_UNITS[columnName]}" data-chart-width="500px" data-chart-height="200px" style="width: 100%; height: 100px;"></canvas>
                        </div>
                    `
                }
            }
            polymorphismDialogContent1 += '</div></div>'
        }

        return polymorphismDialogContent1
    }


    const renderChartsInPolymorphismDialog = (row) => {

        let alleleFrequencyChartData = []
        $('.allele-frequency-chart').each(function() {
            const frequenciesColumn = $(this).data('frequencies-column')
            const frequenciesString = row[frequenciesColumn]
            if (!frequenciesString) return   // Skip if no data

            const frequencies = parseAlleleFrequenciesString(frequenciesString)

            alleleFrequencyChartData.push({
                frequencies: frequencies,
                thisObj: this,
                refAlleleSize: $(this).data('ref-allele-size'),
                axisUnits: $(this).data('axis-units'),
                chartWidth: $(this).data('chart-width'),
                chartHeight: $(this).data('chart-height'),
                label: $(this).data('label'),
                minAlleleSize: Math.min(...frequencies.map(f => f.size)),
                maxAlleleSize: Math.max(...frequencies.map(f => f.size))
            })
        })

        //compute overall min, max for all charts and then create the charts
        let overallMinAlleleSize = Math.min(...alleleFrequencyChartData.map(d => d.minAlleleSize))
        let overallMaxAlleleSize = Math.max(...alleleFrequencyChartData.map(d => d.maxAlleleSize))

        alleleFrequencyChartData.forEach((chartData) => {
            chartData.minAlleleSize = overallMinAlleleSize
            chartData.maxAlleleSize = overallMaxAlleleSize

            createAlleleFrequencyChart(chartData)
        })

        let biallelicHistogramChartData = []
        $('.biallelic-histogram-chart').each(function() {
            const frequenciesColumn = $(this).data('frequencies-column')
            let frequenciesString = row[frequenciesColumn]
            if (!frequenciesString) return   // Skip if no data

            let [data, lookup] = parseBiallelicHistogramString(frequenciesString)

            biallelicHistogramChartData.push({
                data: data,
                lookup: lookup,
                thisObj: this,
                refAlleleSize: $(this).data('ref-allele-size'),
                axisUnits: $(this).data('axis-units'),
                chartWidth: $(this).data('chart-width'),
                chartHeight: $(this).data('chart-height'),
                label: $(this).data('label'),
                minAlleleSize: Math.min(...data.map(f => Math.min(f.x, f.y))),
                maxAlleleSize: Math.max(...data.map(f => Math.max(f.x, f.y)))
            })
        })

        overallMinAlleleSize = Math.min(...biallelicHistogramChartData.map(d => d.minAlleleSize))
        overallMaxAlleleSize = Math.max(...biallelicHistogramChartData.map(d => d.maxAlleleSize))

        biallelicHistogramChartData.forEach((chartData) => {
            chartData.minAlleleSize = overallMinAlleleSize
            chartData.maxAlleleSize = overallMaxAlleleSize

            createBiallelicHistogramChart(chartData)
        })

    }

    const getAlleleHistogramColumnValue = (row, columnName) => {
        if (row[columnName] ) {
            return `<canvas class="allele-frequency-inline-chart cell-in-polymorphism-column" ${getHtmlAttributesForPolymorphism(row)} data-frequencies="${row[columnName]}" data-ref-allele-size="${row['NumRepeatsInReference']}" data-axis-units="${AXIS_UNITS[columnName]}" style="width: 100%; height: 100px"></canvas>`
        } else {
            let otherHistograms = ''
            for ( const c of ALL_STDEV_COLUMNS ) {
                if (row[c] != null) {
                    otherHistograms += `<span class="action-link">σ<sub>${STDEV_SIGMA_SUBSCRIPTS[c]}</sub></span>`
                }
            }

            let result = `<span class="cell-in-polymorphism-column" ${getHtmlAttributesForPolymorphism(row)}><span style="color: #888888; text-align: center; display: block; width: 100%;">`
            if (otherHistograms) {
                result += `not available (but has ${otherHistograms})`
            } else {
                result += `not available`
            }
            result += `</span></span>`
            return result
        }
    }

    let TABLE_COLUMNS = {
        LocusId: {
            displayName: 'LocusId',
            helpText: RESULTS_TABLE_COLUMN_HELP_TEXTS['LocusId'],
            getValue: (row, truncateAt=30) => showTruncatedSequence(row['LocusId'], truncateAt, false),
            getSortColumns: () => ['chrom_index', 'start_0based', 'end_1based', 'LocusId'],
            getRequiredColumns: () => ['chrom_index', 'start_0based', 'end_1based', 'LocusId']
        },
        ReferenceRegion: {
            displayName: 'Interval (hg38)',
            helpText: RESULTS_TABLE_COLUMN_HELP_TEXTS['ReferenceRegion'],
            getValue: (row) => row['ReferenceRegion'].replace(/^chr/i, ''),
            getSortColumns: () => ['chrom_index', 'start_0based', 'end_1based'],
            getRequiredColumns: () => ['chrom_index', 'start_0based', 'end_1based', 'ReferenceRegion']
        },
        ReferenceMotif: {
            displayName: 'Motif',
            helpText: RESULTS_TABLE_COLUMN_HELP_TEXTS['ReferenceMotif'],
            getValue: (row, truncateAt=12) => showTruncatedSequence(row['ReferenceMotif'], truncateAt, true),
            getSortColumns: () => ['MotifSize', 'ReferenceMotif'],
            getRequiredColumns: () => ['MotifSize', 'ReferenceMotif']
        },
        CanonicalMotif: {
            displayName: 'Canonical Motif',
            helpText: RESULTS_TABLE_COLUMN_HELP_TEXTS['CanonicalMotif'],
            getValue: (row, truncateAt=12) => showTruncatedSequence(row['CanonicalMotif'], truncateAt, true),
            getSortColumns: () => ['MotifSize', 'CanonicalMotif'],
            getRequiredColumns: () => ['MotifSize', 'CanonicalMotif']
        },
        ReferenceRepeatSequence: {
            displayName: 'Repeat Sequence',
            helpText: RESULTS_TABLE_COLUMN_HELP_TEXTS['ReferenceRepeatSequence'],
            getValue: (row, truncateAt=50) => showTruncatedSequence(row['ReferenceRepeatSequence'], truncateAt, true),
            getSortColumns: () => ['ReferenceRepeatSequence'],
            getRequiredColumns: () => ['ReferenceRepeatSequence']
        },
        ReferenceLeftFlank: {
            displayName: 'Left Flank',
            helpText: RESULTS_TABLE_COLUMN_HELP_TEXTS['ReferenceLeftFlank'],
            getValue: (row, truncateAt=50) => `<div style='text-align: right'>${showTruncatedSequence(row['ReferenceLeftFlank'], truncateAt, true, true)}</div>`,
            getSortColumns: () => ['ReferenceLeftFlank'],
            getRequiredColumns: () => ['ReferenceLeftFlank']
        },
        ReferenceRightFlank: {
            displayName: 'Right Flank',
            helpText: RESULTS_TABLE_COLUMN_HELP_TEXTS['ReferenceRightFlank'],
            getValue: (row, truncateAt=50) => showTruncatedSequence(row['ReferenceRightFlank'], truncateAt, true),
            getSortColumns: () => ['ReferenceRightFlank'],
            getRequiredColumns: () => ['ReferenceRightFlank']
        },
        GencodeGeneRegion: {
            displayName: 'Region',
            helpText: RESULTS_TABLE_COLUMN_HELP_TEXTS['GencodeGeneRegion'],
            getValue: (row) => getGeneRegionColumnValue(row['GencodeGeneRegion']),
            getSortColumns: () => ['GencodeGeneRegion'],
            getRequiredColumns: () => ['GencodeGeneRegion'],
        },
        GencodeGeneName: {
            displayName: 'Gene',
            defaultSortDirection: 'DESC',
            helpText: RESULTS_TABLE_COLUMN_HELP_TEXTS['GencodeGeneName'],
            getValue: (row) => row['GencodeGeneName'],
            getSortColumns: () => ['GencodeGeneName'],
            getRequiredColumns: () => ['GencodeGeneName'],
        },
        polymorphism: {
            displayName: 'Polymorphism Summary',
            defaultSortDirection: 'DESC',
            helpText: RESULTS_TABLE_COLUMN_HELP_TEXTS['polymorphism'],
            getValue: getPolymorphismSummaryColumnValue,
            getSortColumns: () => ['HPRC100_Stdev'],
            getRequiredColumns: () => REQUIRED_DATA_COLUMNS_FOR_POLYMORPHISM_COLUMN,
        },
        Source: {
            displayName: 'Source',
            helpText: RESULTS_TABLE_COLUMN_HELP_TEXTS['Source'],
            getValue: (row) => {
                const sourceCatalogNameMap = {
                    KnownDiseaseAssociatedLoci: "<span class='show-popup action-link' data-html='<b>Source catalog:</b> TRExplorer v1<br /><br /><b>Original source:</b> Known disease-associated loci'>v1 &nbsp;<i>[1]</i></span>",
                    Illumina174kPolymorphicTRs: "<span class='show-popup action-link' data-html='<b>Source catalog:</b> TRExplorer v1<br /><br /><b>Original source:</b> Illumina 174k polymorphic TR catalog'>v1 &nbsp;<i>[2]</i></span>",
                    PerfectRepeatsInReference:  "<span class='show-popup action-link' data-html='<b>Source catalog:</b> TRExplorer v1<br /><br /><b>Original source:</b> Perfect repeats in hg38 spanning ≥ 9bp and ≥ 3 repeats'>v1 &nbsp;<i>[3]</i></span>",
                    PolymorphicTRsInT2TAssemblies: "<span class='show-popup action-link' data-html='<b>Source catalog:</b> TRExplorer v1<br /><br /><b>Original source:</b> Polymorphic TRs in 78 T2T assemblies'>v1 &nbsp;<i>[4]</i></span>",
                    VamosCatalog: "<span class='show-popup action-link' data-html='<b>Source catalog:</b> TRExplorer v2 draft<br /><br /><b>Original source:</b> Vamos v2.1 catalog'>v2 draft &nbsp; <i>[5]</i></span>",
                }
                const [version, sourceCatalog] = row['Source'].split(':')
                return sourceCatalogNameMap[sourceCatalog]
            },
            getSortColumns: () => ['Source'],
            getRequiredColumns: () => ['Source'],
        },
        AoU1027_OE_LengthPercentile: {
            displayName: 'Constraint (O/E Length) Percentile',
            defaultSortDirection: 'DESC',
            helpText: RESULTS_TABLE_COLUMN_HELP_TEXTS['AoU1027_OE_LengthPercentile'],
            getValue: (row) => row['AoU1027_OE_LengthPercentile'] != null ? `${(100*row['AoU1027_OE_LengthPercentile']).toFixed(1)}%` : '',
            getSortColumns: () => ['AoU1027_OE_LengthPercentile'],
            getRequiredColumns: () => ['AoU1027_OE_LengthPercentile']
        },
        isPathogenic: {
            displayName: 'Disease-associated?',
            defaultSortDirection: 'DESC',
            helpText: RESULTS_TABLE_COLUMN_HELP_TEXTS['isPathogenic'],
            getValue: (row) => row['DiseaseInfo'] ? generateDiseaseAssociatedIcon(row, 'Yes') : '',
            getSortColumns: () => ['DiseaseInfo'],
            getRequiredColumns: () => ['DiseaseInfo']
        },
        variationCluster: {
            displayName: 'VC',
            defaultSortDirection: 'DESC',
            helpText: RESULTS_TABLE_COLUMN_HELP_TEXTS['variationCluster'],
            getValue: (row) => {
                const hasCluster = row['VariationCluster'] && row['VariationCluster'].trim() !== ''
                if (hasCluster) {
                    const refRegionSize = computeIntervalSize(row['ReferenceRegion'])
                    const clusterDetails = `
                            <div class="content" style="text-align: center;">
                                <div><span style="font-weight: bold">Variation Cluster Interval:</span> <br />${row['VariationCluster']}</div>
                                <br />
                                <div><span style="font-weight: bold">Variation Cluster Interval Size:</span> <br /><span style = "white-space: nowrap">reference repeat interval size (${refRegionSize} bp)</span> + ${row['VariationClusterSizeDiff']} bp</div>
                            </div>
                        `.replace(/\s+/g, ' ').trim()

                    return `<span class="action-cell show-popup" data-html='${clusterDetails}'>
                            <i class="check circle icon"></i><span class="action-link">Yes</span>
                        </span>: +${row['VariationClusterSizeDiff']}bp`
                }
                return ''
            },
            getSortColumns: () => ['VariationClusterSizeDiff'],
            getRequiredColumns: () => ['VariationClusterSizeDiff', 'VariationCluster']
        },
        NumRepeatsInReference: {
            displayName: '# of Repeats',
            defaultSortDirection: 'DESC',
            helpText: RESULTS_TABLE_COLUMN_HELP_TEXTS['NumRepeatsInReference'],
            getValue: (row) => (row['NumRepeatsInReference'] ? row['NumRepeatsInReference'] : ''),
            getSortColumns: () => ['NumRepeatsInReference'],
            getRequiredColumns: () => ['NumRepeatsInReference']
        },
        ReferenceRepeatPurity: {
            displayName: 'Purity',
            helpText: RESULTS_TABLE_COLUMN_HELP_TEXTS['ReferenceRepeatPurity'],
            getValue: (row) => (row['ReferenceRepeatPurity'] ? row['ReferenceRepeatPurity'] : ''),
            getSortColumns: () => ['ReferenceRepeatPurity'],
            getRequiredColumns: () => ['ReferenceRepeatPurity']
        },
        TRsInRegion: {
            displayName: 'Nearby TRs',
            defaultSortDirection: 'DESC',
            helpText: RESULTS_TABLE_COLUMN_HELP_TEXTS['TRsInRegion'],
            getValue: (row) => (row['TRsInRegion'] ? (parseInt(row['TRsInRegion']) - 1) : ''),
            getSortColumns: () => ['TRsInRegion'],
            getRequiredColumns: () => ['TRsInRegion']
        },
        RepeatMaskerIntervals: {
            displayName: 'Repeat Masker',
            defaultSortDirection: 'DESC',
            helpText: RESULTS_TABLE_COLUMN_HELP_TEXTS['RepeatMaskerIntervals'],
            getValue: (row) => parseRepeatMaskerIntervals(row['RepeatMaskerIntervals']),
            getSortColumns: () => ['RepeatMaskerIntervals'],
            getRequiredColumns: () => ['RepeatMaskerIntervals'],
        },
        VamosMotifFrequencies: {
            displayName: 'Motif Variability',
            defaultSortDirection: 'DESC',
            helpText: RESULTS_TABLE_COLUMN_HELP_TEXTS['VamosMotifFrequencies'],
            getValue: (row) => parseMotifFrequenciesString(row['VamosMotifFrequencies'], 20),
            getSortColumns: () => ['VamosNumUniqueMotifs', 'VamosMotifFrequencies'],
            getRequiredColumns: () => ['VamosNumUniqueMotifs', 'VamosMotifFrequencies'],
        },
        FlanksAndLocusMappability: {
            displayName: 'Mappability',
            defaultSortDirection: 'DESC',
            helpText: RESULTS_TABLE_COLUMN_HELP_TEXTS['FlanksAndLocusMappability'],
            getValue: (row) => row['FlanksAndLocusMappability'],
            getSortColumns: () => ['FlanksAndLocusMappability'],
            getRequiredColumns: () => ['FlanksAndLocusMappability']
        },
        GencodeGeneId: {
            displayName: 'Gencode Gene Id',
            defaultSortDirection: 'DESC',
            helpText: RESULTS_TABLE_COLUMN_HELP_TEXTS['GencodeGeneId'],
            getValue: (row) => row['GencodeGeneId'],
            getSortColumns: () => ['GencodeGeneId'],
            getRequiredColumns: () => ['GencodeGeneId']
        },
        GencodeTranscriptId: {
            displayName: 'Gencode Transcript',
            defaultSortDirection: 'DESC',
            helpText: RESULTS_TABLE_COLUMN_HELP_TEXTS['GencodeTranscriptId'],
            getValue: (row) => row['GencodeTranscriptId'],
            getSortColumns: () => ['GencodeTranscriptId'],
            getRequiredColumns: () => ['GencodeTranscriptId']
        },
        RefSeqGeneRegion: {
            displayName: 'RefSeq Region',
            defaultSortDirection: 'DESC',
            helpText: RESULTS_TABLE_COLUMN_HELP_TEXTS['RefSeqGeneRegion'],
            getValue: (row) => getGeneRegionColumnValue(row['RefSeqGeneRegion']),
            getSortColumns: () => ['RefSeqGeneRegion'],
            getRequiredColumns: () => ['RefSeqGeneRegion']
        },
        RefSeqGeneId: {
            displayName: 'RefSeq Id',
            defaultSortDirection: 'DESC',
            helpText: RESULTS_TABLE_COLUMN_HELP_TEXTS['RefSeqGeneId'],
            getValue: (row) => row['RefSeqGeneId'],
            getSortColumns: () => ['RefSeqGeneId'],
            getRequiredColumns: () => ['RefSeqGeneId']
        },
        RefSeqTranscriptId: {
            displayName: 'RefSeq Transcript Id',
            defaultSortDirection: 'DESC',
            helpText: RESULTS_TABLE_COLUMN_HELP_TEXTS['RefSeqTranscriptId'],
            getValue: (row) => row['RefseqTranscriptId'],
            getSortColumns: () => ['RefseqTranscriptId'],
            getRequiredColumns: () => ['RefseqTranscriptId']
        },
        RefSeqGeneName: {
            displayName: 'RefSeq Gene',
            defaultSortDirection: 'DESC',
            helpText: RESULTS_TABLE_COLUMN_HELP_TEXTS['RefSeqGeneName'],
            getValue: (row) => row['RefSeqGeneName'],
            getSortColumns: () => ['RefSeqGeneName'],
            getRequiredColumns: () => ['RefSeqGeneName']
        },
        ManeGeneRegion: {
            displayName: 'MANE Region',
            defaultSortDirection: 'DESC',
            helpText: RESULTS_TABLE_COLUMN_HELP_TEXTS['ManeGeneRegion'],
            getValue: (row) => getGeneRegionColumnValue(row['ManeGeneRegion']),
            getSortColumns: () => ['ManeGeneRegion'],
            getRequiredColumns: () => ['ManeGeneRegion']
        },
        ManeGeneId: {
            displayName: 'MANE Gene Id',
            defaultSortDirection: 'DESC',
            helpText: RESULTS_TABLE_COLUMN_HELP_TEXTS['ManeGeneId'],
            getValue: (row) => row['ManeGeneId'],
            getSortColumns: () => ['ManeGeneId'],
            getRequiredColumns: () => ['ManeGeneId']
        },
        ManeTranscriptId: {
            displayName: 'MANE Transcript',
            defaultSortDirection: 'DESC',
            helpText: RESULTS_TABLE_COLUMN_HELP_TEXTS['ManeTranscriptId'],
            getValue: (row) => row['ManeTranscriptId'],
            getSortColumns: () => ['ManeTranscriptId'],
            getRequiredColumns: () => ['ManeTranscriptId']
        },
        ManeGeneName: {
            displayName: 'MANE Gene',
            defaultSortDirection: 'DESC',
            helpText: RESULTS_TABLE_COLUMN_HELP_TEXTS['ManeGeneName'],
            getValue: (row) => row['ManeGeneName'],
            getSortColumns: () => ['ManeGeneName'],
            getRequiredColumns: () => ['ManeGeneName']
        },
        HPRC100_BiallelicHistogram: {
            displayName: 'HPRC100 Biallelic Histogram',
            defaultSortDirection: 'DESC',
            helpText: RESULTS_TABLE_COLUMN_HELP_TEXTS['HPRC100_BiallelicHistogram'],
            getValue: (row) => row['HPRC100_BiallelicHistogram'],
            getSortColumns: () => ['HPRC100_BiallelicHistogram'],
            getRequiredColumns: () => ['HPRC100_BiallelicHistogram']
        },
        TenK10K_BiallelicHistogram: {
            displayName: 'TenK10K Phase 1 Biallelic Histogram',
            defaultSortDirection: 'DESC',
            helpText: RESULTS_TABLE_COLUMN_HELP_TEXTS['TenK10K_BiallelicHistogram'],
            getValue: (row) => row['TenK10K_BiallelicHistogram'],
            getSortColumns: () => ['TenK10K_BiallelicHistogram'],
            getRequiredColumns: () => ['TenK10K_BiallelicHistogram']
        },
        dotPlot: {
            displayName: 'Dot Plot',
            helpText: RESULTS_TABLE_COLUMN_HELP_TEXTS['dotPlot'],
            getValue: (row) => {
                const flankSize = Math.max(24, row["ReferenceMotif"].length * 5)
                const leftFlank = row["ReferenceLeftFlank"].slice(-flankSize)
                const rightFlank = row["ReferenceRightFlank"].slice(0, flankSize)

                return `<canvas class="dotPlot" data-repeat-seq="${row["ReferenceRepeatSequence"]}" data-left-flank="${leftFlank}" data-right-flank="${rightFlank}" data-reference-region="${row['ReferenceRegion']}" data-reference-region-size="${row['ReferenceRegionSize']}" data-flank-width="${flankSize}" data-repeats-in-reference="${row['NumRepeatsInReference']}" data-motif="${row['ReferenceMotif']}" style="width:48px; height:48px"></canvas>`
            },
            getSortColumns: () => ['ReferenceRepeatSequence'],
            getRequiredColumns: () => ['ReferenceRepeatSequence', 'ReferenceLeftFlank', 'ReferenceRightFlank', 'ReferenceRegion', 'NumRepeatsInReference', 'ReferenceMotif', 'ReferenceRegionSize'],
        }
    }


    const REQUIRED_DATA_COLUMNS_FOR_POLYMORPHISM_COLUMN = [
        'ReferenceRegion',
        'NumRepeatsInReference',
        'ReferenceMotif',
        'StdevFromIllumina174k',
        'AlleleFrequenciesFromIllumina174k',
        'StdevFromT2TAssemblies',
        'AlleleFrequenciesFromT2TAssemblies',
        'HPRC100_Stdev',
        'HPRC100_AlleleHistogram',
        'HPRC100_BiallelicHistogram',
        'TenK10K_AlleleHistogram',
        'TenK10K_BiallelicHistogram',
        'TenK10K_Stdev',
        'AoU1027_Stdev',
        'AoU1027_OE_LengthPercentile',
    ]


    const alleleHistogramColumnToOtherColumns = {
        'HPRC100_AlleleHistogram': ['HPRC100_Stdev'],
        'TenK10K_AlleleHistogram': ['TenK10K_Stdev'],
        'AlleleFrequenciesFromIllumina174k': ['StdevFromIllumina174k'],
        'AlleleFrequenciesFromT2TAssemblies': ['StdevFromT2TAssemblies'],
    }


    for (const columnName of ['HPRC100_AlleleHistogram', 'TenK10K_AlleleHistogram', 'AlleleFrequenciesFromIllumina174k', 'AlleleFrequenciesFromT2TAssemblies']) {
        const stdevColumnName = alleleHistogramColumnToOtherColumns[columnName][0]
        TABLE_COLUMNS[columnName] = {
            displayName: `Allele Distribution (${ALLELE_HISTOGRAM_SHORT_NAMES[columnName]})`,
            helpText: ALLELE_HISTOGRAM_HELP_TEXTS[columnName],
            getValue: (row) => getAlleleHistogramColumnValue(row, columnName),
            defaultSortDirection: 'DESC',
            getSortColumns: () => [stdevColumnName, 'NumRepeatsInReference'],
            getRequiredColumns: () => [columnName, 'NumRepeatsInReference', ...REQUIRED_DATA_COLUMNS_FOR_POLYMORPHISM_COLUMN]
        }
    }

    for (const columnName of ['HPRC100_Stdev', 'AoU1027_Stdev', 'TenK10K_Stdev', 'StdevFromIllumina174k', 'StdevFromT2TAssemblies']) {
        TABLE_COLUMNS[columnName] = {
            displayName: `σ<sub>${STDEV_SIGMA_SUBSCRIPTS[columnName]}</sub>`,
            helpText: STDEV_HELP_TEXTS[columnName],
            getValue: (row) => row[columnName],
            defaultSortDirection: 'DESC',
            getSortColumns: () => [columnName],
            getRequiredColumns: () => [columnName]
        }
    }


</script>
</head>

<!-- modal dialog for selecting IGV tracks -->
<div class="ui modal" id="igv-track-selection-dialog">
    <i class="close icon"></i>
    <div class="header" style="white-space: nowrap;">
        Select IGV Tracks
    </div>
    <div class="content" style="padding-top:0px;">
        <div class="ui form">
            <table class="ui stackable table igv-track-selection-table">
                <tr>
                    <td style="padding: 10px 25px !important">
                        <table class="igv-checkbox-container-table" id="igv-tr-explorer-catalog-tracks-checkboxes">
                            <tr>
                                <th class="igv-checkbox-container-header-row">
                                    TRExplorer Catalog
                                </th>
                                <th class="igv-checkbox-container-header-row igv-number-of-loci-column">
                                    # of loci
                                </th>
                            </tr>
                        </table>
                        <table class="igv-checkbox-container-table" id="igv-variation-cluster-tracks-checkboxes">
                            <tr>
                                <th class="igv-checkbox-container-header-row">
                                    Variation Clusters
                                </th>
                                <th class="igv-checkbox-container-header-row igv-number-of-loci-column">
                                </th>
                            </tr>
                        </table>
                        <table class="igv-checkbox-container-table" id="igv-known-disease-loci-checkboxes">
                            <tr>
                                <th class="igv-checkbox-container-header-row">
                                    Known Disease-associated TR Loci
                                </th>
                                <th class="igv-checkbox-container-header-row igv-number-of-loci-column">
                                </th>
                            </tr>
                        </table>
                    </td>
                    <td style="padding: 10px 25px !important">
                        <table class="igv-checkbox-container-table" id="igv-other-tr-catalog-tracks-checkboxes">
                            <tr>
                                <th class="igv-checkbox-container-header-row">
                                    Other Genome-wide TR Catalogs
                                </th>
                                <th class="igv-checkbox-container-header-row igv-number-of-loci-column">
                                    # of loci
                                </th>
                            </tr>
                        </table>
                    </td>
                </tr>
                <tr>
                    <td colspan="2" style="padding: 10px 25px !important">
                        <table class="igv-checkbox-container-table" id="igv-reference-tracks-checkboxes">
                            <tr>
                                <th class="igv-checkbox-container-header-row">Reference Tracks</th>
                            </tr>
                        </table>
                    </td>
                </tr>
                <tr>
                    <td colspan="2" style="padding: 10px 25px !important">
                        <table class="igv-checkbox-container-table" id="igv-custom-tracks" style="width: 100%;">
                            <tr>
                                <th class="igv-checkbox-container-header-row">Custom Tracks <i class="question circle icon link" style="margin-left: 15px;" data-html="These inputs allow you to load your own custom tracks into IGV.js. They accept BED, BAM, VCF, GFF, BigWig, or any other IGV.js-compatible data format. The data file (and index) just have to be online at a publicly-accessible &nbsp;https://, &nbsp;gs://, &nbsp;or s3:// &nbsp; URL."></i></th>
                            </tr>
                        </table>
                    </td>
                </tr>
            </table>
        </div>
    </div>
    <div class="actions">
        <div class="ui deny button">
            Cancel
        </div>
        <div id="igv-tracks-modal-apply-button" class="ui positive right labeled icon button">
            Apply
            <i class="checkmark icon"></i>
        </div>
    </div>
</div>

<div class="ui modal" id="whats-new-modal">
    <i class="close icon"></i>
    <div class="header">
        TRExplorer Updates
    </div>
    <div class="content" style="font-size: 1.1em; height: 100%">
        <iframe src="whats_new.html" style="width: 100%; height: 100%; border: 0px; min-height: 380px"></iframe>
    </div>
</div>


<div class="ui modal" id="about-modal">
    <i class="close icon"></i>
    <div class="header">
        About TRExplorer
    </div>
    <div class="content" style="font-size: 1.1em; height: 100%">
        <iframe src="about.html" style="width: 100%; height: 100%; border: 0px; min-height: 380px"></iframe>
    </div>
</div>


<!-- FAQ modal dialog -->
<div class="ui modal" id="faq-modal" style="height: 80%; min-height:75%">
    <i class="close icon"></i>
    <div class="header">
        Frequently Asked Questions
    </div>
    <div class="content" style="font-size: 1.1em; height: 100%">
        <iframe src="FAQ.html" style="width: 100%; height: 100%; border: 0px"></iframe>
    </div>
</div>

<!-- downloads modal dialog -->
<div class="ui modal" id="downloads-modal" style="height: 88%; min-height: 88%">
    <i class="close icon"></i>
    <div class="header">
        Downloads
    </div>
    <div class="content" style="font-size: 1.1em; height: 100%;">
        <iframe src="downloads.html" style="width: 100%; height: 100%; border: 0px;"></iframe>
    </div>
</div>

